### PR TITLE
Fix array to slice assignment

### LINKED ISSUE #1218: Assigning an array to a slice does't work in Fortran
**Describe the bug**
A wrong Fortran code is generated when trying to assign an array to a slice

**To Reproduce**
For example this python code:
```python
import numpy as np

def func(n:int):
    x = 1
    arr = np.zeros((2, n))
    arr[0:x,0:6:2] = np.array([2,5,6])
    return arr
```

Generate this Fortran code:
```bash/fortran/c
module error_0001


  use, intrinsic :: ISO_C_Binding, only : f64 => C_DOUBLE , i64 => &
        C_INT64_T
  implicit none

  contains

  !........................................
  subroutine func(n, arr) 

    implicit none

    real(f64), allocatable, intent(out) :: arr(:,:)
    integer(i64), value :: n
    integer(i64) :: x

    x = 1_i64
    allocate(arr(0:n - 1_i64, 0:1_i64))
    arr = 0.0_f64
    arr(0_i64:5_i64:2_i64, 0_i64:x - 1_i64) = [2_i64, 5_i64, 6_i64]
    return

  end subroutine func
  !........................................

end module error_0001
```

Which fail in the compilation stage with this error message:
```
22 |     arr(0_i64:5_i64:2_i64, 0_i64:x - 1_i64) = [2_i64, 5_i64, 6_i64]
      |    1
Error: Incompatible ranks 2 and 1 in assignment at (1)
```

**Language**
Fortran

### PR DESCRIPTION
**Pull request by EmilyBourne**
Fix assigning an array to a slice and add tests. Fixes #1218

This is achieved by fixing the indexing in the case where the variable on the left-hand side of the assignment has a higher rank than the variable on the right hand side.

### CODE REVIEW

**Review by Farouk-Echaref — APPROVED** [review]
Looks good!

File: tests/epyccel/modules/arrays.py [hunk 1]
```diff
@@ -1581,6 +1581,29 @@ def array_2d_C_slice_stride_23(a : 'int[:,:]'):
...
+    return arr
+
+def copy_to_slice_1(a : 'float[:]', b : 'float[:]'):
+    a[1:-1] = b
+
+def copy_to_slice_2(a : 'float[:,:]', b : 'float[:]'):
+    a[:,1:-1] = b
+
+def copy_to_slice_3(a : 'float[:,:]', b : 'float[:]'):
+    a[:,0] = b
```
Comment (yguclu) [inline tests/epyccel/modules/arrays.py:1602]: More readable:
```suggestion
def copy_to_slice_3(a : 'float[:,:]', b : 'float[:]'):
    a[:, 0] = b
```

File: tests/epyccel/test_arrays.py [hunk 2]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    epyc_arr = epyc_f(n)
+    check_array_equal(pyth_arr, epyc_arr)
+
+def test_copy_to_slice_1(language):
+    pyth_f = arrays.copy_to_slice_1
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(10, dtype=float)
+    epyc_a = pyth_a.copy()
+    b = np.arange(20,28, dtype=float)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3509]:
```suggestion
    b = np.arange(20, 28, dtype=float)
```

File: tests/epyccel/test_arrays.py [hunk 3]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    epyc_f(epyc_a, b)
+    check_array_equal(pyth_a, epyc_a)
+
+def test_copy_to_slice_2(language):
+    pyth_f = arrays.copy_to_slice_2
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(20, dtype=float).reshape(2,10)
+    epyc_a = pyth_a.copy()
+    b = np.arange(20,28, dtype=float)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3520]:
```suggestion
    b = np.arange(20, 28, dtype=float)
```

File: tests/epyccel/test_arrays.py [hunk 4]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    epyc_f(epyc_a, b)
+    check_array_equal(pyth_a, epyc_a)
+
+def test_copy_to_slice_3(language):
+    pyth_f = arrays.copy_to_slice_3
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(20, dtype=float).reshape(4,5)
+    epyc_a = pyth_a.copy()
+    b = np.arange(20,24, dtype=float)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3531]:
```suggestion
    b = np.arange(20, 24, dtype=float)
```

File: tests/epyccel/test_arrays.py [hunk 5]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    epyc_f(epyc_a, b)
+    check_array_equal(pyth_a, epyc_a)
+
+def test_copy_to_slice_4(language):
+    pyth_f = arrays.copy_to_slice_4
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(10, dtype=float)
+    epyc_a = pyth_a.copy()
+    b = np.arange(20,25, dtype=float)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3542]:
```suggestion
    b = np.arange(20, 25, dtype=float)
```

File: tests/epyccel/test_arrays.py [hunk 6]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    b = np.arange(20,28, dtype=float)
+    pyth_f(pyth_a, b)
+    epyc_f(epyc_a, b)
+    check_array_equal(pyth_a, epyc_a)
+
+def test_copy_to_slice_2(language):
+    pyth_f = arrays.copy_to_slice_2
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(20, dtype=float).reshape(2,10)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3518]:
```suggestion
    pyth_a = np.arange(20, dtype=float).reshape(2, 10)
```

File: tests/epyccel/test_arrays.py [hunk 7]
```diff
@@ -3488,6 +3488,62 @@ def test_array_2d_C_slice_stride_23(language):
...
+    b = np.arange(20,28, dtype=float)
+    pyth_f(pyth_a, b)
+    epyc_f(epyc_a, b)
+    check_array_equal(pyth_a, epyc_a)
+
+def test_copy_to_slice_3(language):
+    pyth_f = arrays.copy_to_slice_3
+    epyc_f = epyccel(pyth_f, language = language)
+
+    pyth_a = np.arange(20, dtype=float).reshape(4,5)
```
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3529]:
```suggestion
    pyth_a = np.arange(20, dtype=float).reshape(4, 5)
```

File: pyccel/ast/utilities.py [hunk 8]
```diff
@@ -541,22 +541,34 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
...
-                          for j, rj in enumerate(rhs)]
-            collect_loops(new_assigns, indices, new_index, language_has_vectors, result = result)
+            if lhs.rank > rhs.rank:
+                for index in range(lhs.rank-rhs.rank):
+                    # If an index exists at the same depth, reuse it if not create one
+                    if index >= len(indices):
+                        indices.append(new_index(PythonNativeInt(),'i'))
+                    index_var = indices[index]
+                    lhs = insert_index(lhs, index, index_var)
+                collect_loops([Assign(lhs,rhs)], indices, new_index, language_has_vectors, result = result)
```
Comment (yguclu) [inline pyccel/ast/utilities.py:554]: The code is more readable if we follow a comma with an empty space. There are a couple missing here:
```suggestion
                        indices.append(new_index(PythonNativeInt(), 'i'))
                    index_var = indices[index]
                    lhs = insert_index(lhs, index, index_var)
                collect_loops([Assign(lhs, rhs)], indices, new_index, language_has_vectors, result = result)
```

File: pyccel/ast/utilities.py [hunk 9]
```diff
@@ -541,22 +541,34 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
...
-                            rj) # lhs[j] = rhs[j]
-                          for j, rj in enumerate(rhs)]
-            collect_loops(new_assigns, indices, new_index, language_has_vectors, result = result)
+            if lhs.rank > rhs.rank:
+                for index in range(lhs.rank-rhs.rank):
+                    # If an index exists at the same depth, reuse it if not create one
+                    if index >= len(indices):
+                        indices.append(new_index(PythonNativeInt(),'i'))
+                    index_var = indices[index]
+                    lhs = insert_index(lhs, index, index_var)
```
Comment (yguclu) [inline pyccel/ast/utilities.py:553]: I am a bit confused by the names of the variables here. `index` is an `int` and `index_var` is a `Variable`, but `indices` is a list of `Variable` objects. Either we rename `indices` as `var_indices` (and probably `index_var` as `var_index`, for consistency), or we rename (`index`, `index_var`) as (`index_int`, `index`).

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good to me. I have some minor comments as usual.

File: pyccel/ast/utilities.py (hunk 9 above)
Comment (EmilyBourne) [reply pyccel/ast/utilities.py]: I have used the surrounding comments for inspiration and have renamed (`index`,`index_var`) to (`index_depth`, `index`) : bef7be94b

File: pyccel/ast/utilities.py [hunk 10]
```diff
@@ -494,14 +494,14 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
             # Loop over indexes, inserting until the expression can be evaluated
             # in the desired language
             new_level = 0
-            for index in range(-rank,0):
+            for index_depth in range(-rank,0):
```
Comment (yguclu) [inline pyccel/ast/utilities.py:497]: `index_depth` is a great name!

By the way:
```suggestion
            for index_depth in range(-rank, 0):
```

File: pyccel/ast/utilities.py [hunk 11]
```diff
@@ -494,14 +494,14 @@ def collect_loops(block, indices, new_index, language_has_vectors = False, resul
...
                 new_level += 1
                 # If an index exists at the same depth, reuse it if not create one
-                if rank+index >= len(indices):
+                if rank+index_depth >= len(indices):
                     indices.append(new_index(PythonNativeInt(),'i'))
```
Comment (yguclu) [inline pyccel/ast/utilities.py:501]:
```suggestion
                    indices.append(new_index(PythonNativeInt(), 'i'))
```

**Review by yguclu — CHANGES_REQUESTED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**